crates/ide_assists/src/handlers/inline_type_alias.rs

```diff
@@ -0,0 +1,714 @@
+/// versa if they're in the wrong position. It supports partially specified
+/// generics.
+///
+/// 1. Map the provided instance's generic args to the type alias's generic
+///    params:
+///
+///    ```
+///    type A<'a, const N: usize, T = u64> = &'a [T; N];
+///          ^ alias generic params
+///    let a: A<100>;
+///            ^ instance generic args
+///    ```
+///
+///    generic['a] = '_ due to omission
+///    generic[N] = 100 due to the instance arg
+///    generic[T] = u64 due to the default param
+///
+/// 2. Copy the concrete type and substitute in each found mapping:
+///
+///    &'_ [u64; 100]
+///
+/// 3. Remove wildcard lifetimes entirely:
+///
+///    &[u64; 100]
+fn get_replacement_for_generic_alias(
+    instance_generic_args_list: Option<ast::GenericArgList>,
+    alias_generics: ast::GenericParamList,
+    concrete_type: &ast::Type,
+) -> Option<String> {
+    if alias_generics.generic_params().count() == 0 {
+        cov_mark::hit!(no_generics_params);
+        return None;
+    }
+
+    let mut lifetime_mappings = HashMap::<&str, ast::Lifetime>::new();
+    let mut other_mappings = HashMap::<String, SyntaxNode>::new();
+
+    let wildcard_lifetime = make::lifetime("'_");
+    let alias_lifetimes = alias_generics.lifetime_params().map(|l| l.to_string()).collect_vec();
+    for lifetime in &alias_lifetimes {
+        lifetime_mappings.insert(lifetime, wildcard_lifetime.clone());
+    }
+
+    if let Some(ref instance_generic_args_list) = instance_generic_args_list {
+        for (index, lifetime) in instance_generic_args_list
+            .lifetime_args()
+            .map(|arg| arg.lifetime().expect("LifetimeArg has a Lifetime"))
+            .enumerate()
+        {
+            if index >= alias_lifetimes.len() {
+                cov_mark::hit!(too_many_lifetimes);
+                return None;
+            }
+
+            let key = &alias_lifetimes[index];
+
+            lifetime_mappings.insert(key, lifetime);
+        }
+    }
+
+    let instance_generics = generic_args_to_other_generics(instance_generic_args_list);
+    let alias_generics = generic_param_list_to_other_generics(&alias_generics);
+
+    if instance_generics.len() > alias_generics.len() {
+        cov_mark::hit!(too_many_generic_args);
+        return None;
+    }
+
+    // Any declaration generics that don't have a default value must have one
+    // provided by the instance.
+    for (i, declaration_generic) in alias_generics.iter().enumerate() {
+        let key = declaration_generic.replacement_key();
+
+        if let Some(instance_generic) = instance_generics.get(i) {
+            other_mappings.insert(key, instance_generic.replacement_value()?);
+        } else if let Some(value) = declaration_generic.replacement_value() {
+            other_mappings.insert(key, value);
+        } else {
+            cov_mark::hit!(missing_replacement_param);
+            return None;
+        }
+    }
+
```

Comment:
From this point on `get_replacement_for_generic_alias` can't fail anymore, so I think it would make sense to split it in mapping creation and replacement application. Then we can do the replacement application inside the builder closure (which only runs when the assist is being executed) making the assist calculation cheaper.